contracts/SpeedMarkets/ChainedSpeedMarketsAMM.sol

```diff
@@ -0,0 +1,521 @@
+    }
+
+    function _getPayout(uint buyinAmount, uint numOfDirections) internal returns (uint payout) {
+        payout = buyinAmount;
+        for (uint i = 0; i < numOfDirections; i++) {
+            payout = (payout * payoutMultiplier) / ONE;
+        }
+    }
+
+    function _createNewMarket(
+        bytes32 asset,
+        uint64 timeFrame,
+        SpeedMarket.Direction[] calldata directions,
+        uint buyinAmount,
+        bytes[] memory priceUpdateData,
+        bool transferSusd,
+        address referrer
+    ) internal {
+        if (referrer != address(0)) {
+            speedMarketsAMM.referrals().setReferrer(referrer, msg.sender);
+        }
+        require(speedMarketsAMM.supportedAsset(asset), "Asset is not supported");
+        require(buyinAmount >= minBuyinAmount && buyinAmount <= maxBuyinAmount, "Wrong buy in amount");
+        require(timeFrame >= minTimeFrame && timeFrame <= maxTimeFrame, "Wrong time frame");
+        require(
+            directions.length >= minChainedMarkets && directions.length <= maxChainedMarkets,
+            "Wrong number of directions"
+        );
+        require(_getPayout(buyinAmount, directions.length) <= maxProfitPerIndividualMarket, "Profit too high");
+
+        currentRisk +=
+            _getPayout(buyinAmount, directions.length) -
+            (buyinAmount * (ONE + speedMarketsAMM.safeBoxImpact())) /
+            ONE;
+        require(currentRisk <= maxRisk, "Out of liquidity");
+
+        speedMarketsAMM.pyth().updatePriceFeeds{value: speedMarketsAMM.pyth().getUpdateFee(priceUpdateData)}(
+            priceUpdateData
+        );
+
+        PythStructs.Price memory price = speedMarketsAMM.pyth().getPrice(speedMarketsAMM.assetToPythId(asset));
+        require(
+            (price.publishTime + speedMarketsAMM.maximumPriceDelay()) > block.timestamp && price.price > 0,
+            "Stale price"
+        );
+
+        if (transferSusd) {
+            uint totalAmountToTransfer = (buyinAmount * (ONE + speedMarketsAMM.safeBoxImpact())) / ONE;
+            sUSD.safeTransferFrom(msg.sender, address(this), totalAmountToTransfer);
+        }
+
+        ChainedSpeedMarket csm = ChainedSpeedMarket(Clones.clone(chainedSpeedMarketMastercopy));
+        csm.initialize(
+            ChainedSpeedMarket.InitParams(
+                address(this),
+                msg.sender,
+                asset,
+                timeFrame,
+                uint64(block.timestamp + timeFrame),
+                uint64(block.timestamp + timeFrame * directions.length), // strike time
+                price.price,
+                directions,
+                buyinAmount,
+                speedMarketsAMM.safeBoxImpact()
+            )
+        );
+
+        sUSD.safeTransfer(address(csm), _getPayout(buyinAmount, directions.length));
+
+        {
+            uint referrerShare;
+            if (address(speedMarketsAMM.referrals()) != address(0)) {
```

Comment:
multiple reads of the same getter from another contract are expensive. Please read once and store it in a local variable.

Additionally, this might be a good chance to create a contract called AddressManager that would have the methods to get Referrer contract, SafeBox contract, Pyth contract, etc... and be reused from other contracts. Rather than each AMM having its own variable for Referral contract address.